app/join_record.ts

```diff
@@ -0,0 +1,127 @@
+"use server";
+// import { access, constants, appendFileSync, readFile } from "node:fs";
+import {
+  S3Client,
+  PutObjectCommand,
+  GetObjectCommand,
+} from "@aws-sdk/client-s3";
+import { JoinFormValues } from "@/components/JoinForm/JoinForm";
+import { Readable } from "stream";
+import { JoinRecord } from "./types";
+
+// const JOIN_FORM_LOG = process.env.JOIN_FORM_LOG as string;
+
+class JoinRecordS3 {
+  private s3Client: S3Client;
+
+  private S3_REGION: string;
+  private S3_ENDPOINT: string;
+  private S3_BUCKET_NAME: string;
+  private S3_BASE_NAME: string;
+  private S3_ACCESS_KEY: string;
+  private S3_SECRET_KEY: string;
+
+  constructor() {
+    this.S3_REGION = process.env.S3_REGION as string;
+    this.S3_ENDPOINT = process.env.S3_ENDPOINT as string;
+    this.S3_BUCKET_NAME = process.env.S3_BUCKET_NAME as string;
+    this.S3_BASE_NAME = process.env.S3_BASE_NAME as string;
+    this.S3_ACCESS_KEY = process.env.S3_ACCESS_KEY as string;
+    this.S3_SECRET_KEY = process.env.S3_SECRET_KEY as string;
+
+    // Setup the S3 client
+    this.s3Client = new S3Client({
+      region: this.S3_REGION != undefined ? this.S3_REGION : "us-east-1",
+      endpoint:
+        this.S3_ENDPOINT != undefined
+          ? this.S3_ENDPOINT
+          : "https://s3.us-east-1.amazonaws.com",
+      credentials: {
+        accessKeyId: this.S3_ACCESS_KEY,
+        secretAccessKey: this.S3_SECRET_KEY,
+      },
+    });
+  }
+
+  // Records the submission we just got as a JSON object in an S3 bucket.
+  // submission: A Join Form Submission. We append a few things to this.
+  // key: The S3 path we store the submission at
+  // responseCode: If we have a response code for this submission, add it here.
+  async save(joinRecord: JoinRecord, key: string = "") {
+    // Bail if there's no S3 key
+    if (this.S3_ACCESS_KEY === undefined || this.S3_SECRET_KEY === undefined) {
```

Comment:
If you take my advice above, I think this could become:
```suggestion
    if (this.s3Client.config.credentials.accessKeyId === undefined || this.s3Client.config.credentials.secretAccessKey === undefined) {
```

Reply:
My IDE doesn't like that, but I found another way that seems to work.